File: java-mvc/clean-architecture/src/main/java/com/ecommerce/clean/usecase/UpdateProductUseCase.java
package com.ecommerce.clean.usecase;

import com.ecommerce.clean.entity.Product;
import com.ecommerce.clean.usecase.port.ProductGateway;

import java.math.BigDecimal;
import java.util.UUID;

/**
 * USE CASE: Update product details.
 */
public class UpdateProductUseCase {

    private final ProductGateway productGateway;

    public UpdateProductUseCase(ProductGateway productGateway) {
        this.productGateway = productGateway;
    }

    public Product execute(UUID id, String name, String description, BigDecimal price, String category) {
        Product product = productGateway.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Product not found: " + id));

        product.updateDetails(name, description, price, category);
        return productGateway.save(product);
    }
}

File: nodejs-mvc/clean-architecture/src/usecases/UpdateProductUseCase.ts
import { Product } from '../entities/Product';
import { ProductGateway } from './interfaces/ProductGateway';
import { ProductOutput, UpdateProductInput } from './dto/ProductDTO';
import { DuplicateSkuError, ProductNotFoundError } from './errors/UseCaseErrors';

/**
 * @layer Use Cases
 * @description Updates an existing product's details.
 */
export class UpdateProductUseCase {
  constructor(private readonly gateway: ProductGateway) {}

  async execute(id: string, input: UpdateProductInput): Promise<ProductOutput> {
    const existing = await this.gateway.findById(id);
    if (!existing) {
      throw new ProductNotFoundError(id);
    }

    // Check SKU uniqueness if changing
    if (input.sku && input.sku !== existing.sku) {
      const bySku = await this.gateway.findBySku(input.sku);
      if (bySku) {
        throw new DuplicateSkuError(input.sku);
      }
    }

    // Entity handles validation of business rules
    const updated = existing.updateDetails(input);
    const saved = await this.gateway.update(updated);
    return this.toOutput(saved);
  }

  private toOutput(product: Product): ProductOutput {
    return {
      id: product.id,
      name: product.name,
      description: product.description,
      price: product.price,
      category: product.category,
      stockQuantity: product.stockQuantity,
      sku: product.sku,
      active: product.active,
      createdAt: product.createdAt.toISOString(),
      updatedAt: product.updatedAt.toISOString(),
    };
  }
}
